File: src/app/useCases/trafficSource/listTrafficSourceByIdUseCase.ts
import { UserGatewayDTO } from "../../../domain/gateways/user";
import { TrafficSourceRepository } from "../../../domain/repositories/trafficSource";
import { HttpAdapter } from "../../../infra/adapters/httpAdapter";

class ListTrafficSourceByIdUseCase {
  constructor(
    private trafficSourceRepository: TrafficSourceRepository,
    private userGateway: UserGatewayDTO
  ) {}

  async execute(trafficSourceId: string, token: string) {
    const [trafficSource, user] = await Promise.all([
      this.trafficSourceRepository.findById(trafficSourceId),
      this.userGateway.findUnique(token),
    ]);

    if (!trafficSource) throw HttpAdapter.notFound("Traffic Source not found");
    return trafficSource.toJson(user.utc);
  }
}

export { ListTrafficSourceByIdUseCase };

File: internal/app/useCases/trafficSource/listTrafficSourceByIdUseCase.go
package trafficSourceUseCases

import (
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/entities"
	domainErrors "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/errors"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/repositories"
)

type ListTrafficSourceByIdUseCase struct {
	trafficSourceRepository repositories.TrafficSourceRepository
}

func NewListTrafficSourceByIdUseCase(trafficSourceRepository repositories.TrafficSourceRepository) *ListTrafficSourceByIdUseCase {
	return &ListTrafficSourceByIdUseCase{trafficSourceRepository: trafficSourceRepository}
}

func (t *ListTrafficSourceByIdUseCase) Handle(trafficSourceId string) (*entities.TrafficSource, error) {
	trafficSource, err := t.trafficSourceRepository.FindById(trafficSourceId)

	if err != nil {
		return nil, err
	}

	if trafficSource == nil {
		return nil, domainErrors.NotFound("Traffic source not found")
	}

	return trafficSource, nil
}
